===========================================Tiền Xử Lí Và Khám Phá Dữ Liệu=================================

1. IMPORT THƯ VIỆN

In [1]:

import numpy as np
import pandas as pd

2. ĐỌC DỮ LIỆU KIỂM TRA VÀ KIỂM TRA SỐ DÒNG, SỐ CỘT DỮ LIỆU

In [2]:
df = pd.read_csv("../Data/data.csv")
df.head()

Country  Year Fossil fuels         Gas        Coal         Oil  Solar  \
0     Algeria  2000      300 TWh  191.01 TWh    5.64 TWh  103.13 TWh  0 TWh   
1   Argentina  2000      595 TWh  322.97 TWh    7.51 TWh  264.94 TWh  0 TWh   
2   Australia  2000    1,253 TWh  205.55 TWh  591.66 TWh  455.98 TWh  0 TWh   
3     Austria  2000      262 TWh   78.21 TWh   42.47 TWh  141.75 TWh  0 TWh   
4  Azerbaijan  2000      128 TWh   52.83 TWh    0.00 TWh   74.69 TWh  0 TWh   

    Wind Renewables Hydropower  
0  0 TWh      0 TWh      0 TWh  
1  0 TWh     86 TWh     84 TWh  
2  0 TWh     52 TWh     49 TWh  
3  0 TWh    129 TWh    124 TWh  
4  0 TWh      5 TWh      5 TWh

In [3]:
shape = df.shape
print(f"Shape: {shape}")

Shape: (1840, 10)


In [4]:
#Số dòng
if shape[0] > 1000:
    print(f"Số dòng dữ liệu đạt yêu cầu.")
else:
    print(f"Số dòng dữ liệu không đạt yêu cầu")
#Số cột
if shape[1] >= 5:
    print(f"Số cột dữ liệu đạt yêu cầu.")
else:
    print(f"Số cột dữ liệu không đạt yêu cầu")
#Kiểm tra có dòng duplicate không
num_duplicated_rows = df.index.duplicated().sum()
if num_duplicated_rows == 0:
    print(f"Dữ liệu không có dòng lặp lại.")
else:
    print(f"Dữ liệu bạn có {num_duplicated_rows} lặp lại.")

Số dòng dữ liệu đạt yêu cầu.
Số cột dữ liệu đạt yêu cầu.
Dữ liệu không có dòng lặp lại.


3. Ý NGHĨA CỦA CÁC CỘT TRONG DATAFRAME VÀ THÔNG TIN VỀ DATAFRAME

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1840 entries, 0 to 1839
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Country       1840 non-null   object
 1   Year          1840 non-null   int64 
 2   Fossil fuels  1697 non-null   object
 3   Gas           1752 non-null   object
 4   Coal          1762 non-null   object
 5   Oil           1817 non-null   object
 6   Solar         1805 non-null   object
 7   Wind          1817 non-null   object
 8   Renewables    1639 non-null   object
 9   Hydropower    1639 non-null   object
dtypes: int64(1), object(9)
memory usage: 143.9+ KB


- Ý nghĩa về các cột trong DataFrame:
    **Chú ý:** các đơn vị năng lượng sử dụng ở đây là TWh, vì đây là lượng năng lượng sử dụng cả một năm.
    + Country: Tên quốc gia mà chúng ta đang đề cập đến.
    + Year: Năm tương ứng.
    Các cột còn lại là tổng năng lượng tiêu thụ từ các nguồn của quốc gia đó trong năm (Year), bao gồm:
    + Fosil fuels (bao gồm cả Coal, Oil, và Gas): Năng lượng hóa thạch
    + Gas: Năng lượng từ khí đốt
    + Coal: Năng lượng từ than đá
    + Oil: Năng lượng từ dầu
    + Solar: Năng lượng mặt trời
    + Wind: Năng lượng gió
    + Renewables: năng lượng tái tạo
    + Hydropower: năng lượng từ thủy điện

- Tổng quan về dữ liệu:
    + Dữ liệu có 1840 dòng và 9 cột.
    + Có 1 đối tượng kiểu in64 là Year, còn lại là kiểu object
    + Số lượng dữ liệu không phải null thể hiện như ở trên.

4. XỬ LÝ DỮ LIỆU, CHUYỂN CÁC KIỂU DỮ LIỆU CẦN THIẾT VỀ NUMERIC VÀ XEM XÉT BỎ CÁC CỘT CÓ ĐỘ MẤT MÁT DỮ LIỆU LỚN


        4.1 Xử lý dữ liệu, chuyển các kiểu dữ liệu về kiểu numeric

In [6]:
#Chuyển các dữ liệu về kiểu số.
convert_columns=df.columns[2:]#Bỏ ra cột tên nước và năm đi
for col in convert_columns:
    df[col] = pd.to_numeric(df[col].str.replace(' TWh', '').str.replace(',', ''),errors='coerce')
#Thêm đơn vị vào tên cột
df = df.add_suffix(" (TWh)")
df.rename(columns = {'Country (TWh)':'Country', 'Year (TWh)':'Year'}, inplace = True) 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1840 entries, 0 to 1839
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Country             1840 non-null   object 
 1   Year                1840 non-null   int64  
 2   Fossil fuels (TWh)  1697 non-null   float64
 3   Gas (TWh)           1752 non-null   float64
 4   Coal (TWh)          1762 non-null   float64
 5   Oil (TWh)           1817 non-null   float64
 6   Solar (TWh)         1805 non-null   float64
 7   Wind (TWh)          1817 non-null   float64
 8   Renewables (TWh)    1639 non-null   float64
 9   Hydropower (TWh)    1639 non-null   float64
dtypes: float64(8), int64(1), object(1)
memory usage: 143.9+ KB


- Sau khi thực hiện như trên ta thấy các dữ liệu các dữ liệu năng lượng đều chuyển về kiểu float64.

        4.2 Tìm kiếm các cột dữ liệu bị mất quá nhiều.

In [7]:
#Tính toán giá trị phần trăm mà dữ liệu bị mất và hiển thị thêm thông tin về giá trị lớn nhất và nhỏ nhất.
raw_df = df.select_dtypes(exclude=['object']) # chỉ lấy kiểu numeric, không lấy kiểu object,str
def missing_ratio(raw_df):
    return raw_df.isna().mean() * 100
num_col_info_df = raw_df.agg([missing_ratio, "min", "mean", "max"]).round(1)
num_col_info_df

Year  Fossil fuels (TWh)  Gas (TWh)  Coal (TWh)  Oil (TWh)  \
missing_ratio     0.0                 7.8        4.8         4.2        1.2   
min            2000.0                12.0        0.0         0.0        7.2   
mean           2011.0              1570.7      411.3       511.6      584.9   
max            2022.0             36319.0     8812.1     24559.5    11214.1   

               Solar (TWh)  Wind (TWh)  Renewables (TWh)  Hydropower (TWh)  
missing_ratio          1.9         1.2              10.9              10.9  
min                    0.0         0.0               0.0               0.0  
mean                   8.9        21.9             184.7             123.4  
max                 1115.0      1988.0            7093.0            3471.0

- Ở bảng trên ta thấy dòng đầu tiên là tỉ lệ phần trăm dữ liệu bị mất, tiếp theo là các dòng giá trị nhỏ nhất, trung bình và lớn nhất của tương ứng với từng cột thuộc tính.
- Cột year ở đây không có ý nghĩa nhưng mà ta thêm vào để thấy được dữ liệu bị mất của nó là 0% chứng tỏ chúng ta đang làm đúng.

        4.3 Xử lí nếu các cột dữ liệu bị mất quá nhiều

In [8]:
def drop_missing_features(df: pd.DataFrame, missing_lst: dict = dict(num_col_info_df.iloc[0]), threshold: float = 75.0) -> pd.DataFrame:
    # YOUR CODE HERE
    columns_to_drop = [col for col, missing_rate in missing_lst.items() if missing_rate >= threshold]

    # Drop the columns from the DataFrame.
    df_dropped = df.drop(columns=columns_to_drop)

    return df_dropped

In [9]:
#Test
raw_df = drop_missing_features(df, num_col_info_df.iloc[0], threshold = 75.0)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1840 entries, 0 to 1839
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Country             1840 non-null   object 
 1   Year                1840 non-null   int64  
 2   Fossil fuels (TWh)  1697 non-null   float64
 3   Gas (TWh)           1752 non-null   float64
 4   Coal (TWh)          1762 non-null   float64
 5   Oil (TWh)           1817 non-null   float64
 6   Solar (TWh)         1805 non-null   float64
 7   Wind (TWh)          1817 non-null   float64
 8   Renewables (TWh)    1639 non-null   float64
 9   Hydropower (TWh)    1639 non-null   float64
dtypes: float64(8), int64(1), object(1)
memory usage: 143.9+ KB


        4.4 Điền các giá trị thiếu bằng trung bình các giá trị của cột

In [10]:
def filling_missing_value(df: pd.DataFrame) -> pd.DataFrame:
    # YOUR CODE HERE
    # Replace all None values with NaN
    df = df.replace({None: np.nan, 0: np.nan})

    # Fill NaN values with the mean of the column values
    for i in df.columns[df.isnull().any(axis=0)]:    
        df[i].fillna(df[i].mean().round(1),inplace=True)
    return df

In [11]:
#Test
df = filling_missing_value(df)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1840 entries, 0 to 1839
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Country             1840 non-null   object 
 1   Year                1840 non-null   int64  
 2   Fossil fuels (TWh)  1840 non-null   float64
 3   Gas (TWh)           1840 non-null   float64
 4   Coal (TWh)          1840 non-null   float64
 5   Oil (TWh)           1840 non-null   float64
 6   Solar (TWh)         1840 non-null   float64
 7   Wind (TWh)          1840 non-null   float64
 8   Renewables (TWh)    1840 non-null   float64
 9   Hydropower (TWh)    1840 non-null   float64
dtypes: float64(8), int64(1), object(1)
memory usage: 143.9+ KB


In [12]:
#check data
df.head()

Country  Year  Fossil fuels (TWh)  Gas (TWh)  Coal (TWh)  Oil (TWh)  \
0     Algeria  2000               300.0     191.01        5.64     103.13   
1   Argentina  2000               595.0     322.97        7.51     264.94   
2   Australia  2000              1253.0     205.55      591.66     455.98   
3     Austria  2000               262.0      78.21       42.47     141.75   
4  Azerbaijan  2000               128.0      52.83      528.10      74.69   

   Solar (TWh)  Wind (TWh)  Renewables (TWh)  Hydropower (TWh)  
0         27.7        46.0             194.4             140.5  
1         27.7        46.0              86.0              84.0  
2         27.7        46.0              52.0              49.0  
3         27.7        46.0             129.0             124.0  
4         27.7        46.0               5.0               5.0

5. Sự phân bổ các giá trị

    5.1  Với các cột có kiểu dữ liệu không phải số

In [13]:
pd.set_option('display.max_colwidth', 100) # For clearly
pd.set_option('display.max_columns', None) # For clearly

In [14]:
df["Year"] = df["Year"].astype("str")
cat_col_info_df = df.select_dtypes(include='object')

def missing_ratio(df):
    return (df.isna().mean() * 100).round(1)

def num_values(col):
    if col.name not in cat_col_info_df.columns:
        return col.nunique()
    else:
        s = col.str.split(';').explode()
        return s.nunique()
    
def value_ratios(col):
    if col.name not in cat_col_info_df.columns:
        return dict((col.value_counts(normalize=True)*100).round(1))
    else:
        s = col.str.split(';').explode()
        s = (s.value_counts(normalize=True, ascending=False)*100).round(1)
        return dict(s)   

cat_col_info_df = cat_col_info_df.agg([missing_ratio, num_values, value_ratios])
cat_col_info_df

Country  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                      80   
value_ratios   {'Algeria': 1.2, 'Argentina': 1.2, 'Singapore': 1.2, 'Saudi Arabia': 1.2, 'Russia': 1.2, 'Romani...   

                                                                                                              Year  
missing_ratio                                                                                                  0.0  
num_values                                                                                                      23  
value_ratios   {'2000': 4.3, '2012': 4.3, '2021': 4.3, '2020': 4.3, '2019': 4.3, '2018': 4.3, '2017': 4.3, '201...

    5.2 Với các cột có kiểu dữ liệu số

In [15]:
# YOUR CODE HERE
num_col_info_df = df.select_dtypes(include = 'number')
def missing_ratio(df):
    return df.isna().mean() * 100
def lower_quartile(df):
    return df.quantile(0.25)
def median(df):
    return df.quantile(0.5)
def upper_quartile(df):
    return df.quantile(0.75)
num_col_info_df = num_col_info_df.agg([missing_ratio, "min", lower_quartile, median, upper_quartile, "max"]).round(1)
df.select_dtypes(exclude='object').agg([missing_ratio, "min", lower_quartile, median, upper_quartile, "max"])

Fossil fuels (TWh)  Gas (TWh)  Coal (TWh)   Oil (TWh)  \
missing_ratio                 0.00     0.0000      0.0000      0.0000   
min                          12.00     0.0800      0.0100      7.2000   
lower_quartile              207.75    43.5375     10.1100     85.4025   
median                      551.50   168.5700     54.2500    175.1150   
upper_quartile             1570.70   412.2000    365.7325    563.7700   
max                       36319.00  8812.1200  24559.4900  11214.0800   

                Solar (TWh)  Wind (TWh)  Renewables (TWh)  Hydropower (TWh)  
missing_ratio           0.0        0.00               0.0               0.0  
min                     1.0        1.00               1.0               1.0  
lower_quartile         23.0        9.75              19.0              14.0  
median                 27.7       46.00              72.5              62.5  
upper_quartile         27.7       46.00             194.4             140.5  
max                  1115.0     1988.00            7093.0            3471.0

6. Kiểm tra tính hợp lệ của dữ liệu

In [16]:
def checking_resonable(df: pd.DataFrame) -> bool:
    for col in df.columns:
        if df[col].min() < 0 :
            return False
    return True
#check
raw_df = df.select_dtypes(include = 'number')
checking_resonable(raw_df)

True

In [17]:
# Lưu file đã xử lý
df.to_csv("../Data/data__processed.csv")